### PR TITLE
return edgeptr_t in layer getters

### PR DESCRIPTION
**Pull request by edgarriba**
Not compiling yet. Please don't merge

### CODE REVIEW

File: tiny_cnn/layers/layer.h [hunk 1]
```diff
@@ -496,21 +496,24 @@ class layer : public node {
...
-        return dynamic_cast<edge*>(prev()[i].get());
+        return prev()[i];
     }
-    const edge* ith_in_node(cnn_size_t i) const {
+
+    const edgeptr_t ith_in_node(cnn_size_t i) const {
         if (!prev_[i]) alloc_input(i);
-        return dynamic_cast<const edge*>(prev()[i].get());
+        return prev()[i];
     }
```
Comment (edgarriba) [inline tiny_cnn/layers/layer.h]: @nyanp got big compilation errors here. Any idea?

File: tiny_cnn/layers/layer.h [hunk 2]
```diff
@@ -496,21 +496,24 @@ class layer : public node {
...
+        return next()[i];
     }
-    const edge* ith_out_node(cnn_size_t i) const {
-        if (!next_[i]) alloc_output(i);
-        return dynamic_cast<const edge*>(next()[i].get());
+
+    const edgeptr_t ith_out_node(cnn_size_t i) const {
+        if (!next_[i]) alloc_input(i);
+        return next()[i];
     }
```
Comment (edgarriba) [inline tiny_cnn/layers/layer.h]: @nyanp got big compilation errors here due to constantness property. Any idea?

**Comment by nyanp** [comment]
@edgarriba 
- c++ standard doesn't allow ```vector<const T>```, so use ```vector<edgeptr_t>``` instead of ```vector<const edgeptr_t>```. 

 > note: ```vector<const T*>``` is valid, because a variable of ```const T*``` is copyable but ```const T``` is not. This is why you can compile ```vector<const edge*>``` but can't compile ```vector<const edgeptr_t>```. see also:[http://stackoverflow.com/questions/6954906/does-c11-allow-vectorconst-t](http://stackoverflow.com/questions/6954906/does-c11-allow-vectorconst-t))

- we can also declare the returning type of ```ith_in/out_node() const``` as ```edgeptr_t```, instead of ```const edgeptr_t```.
- after that, we can favorably remove non-const overload of ```ith_in/out_node()``` methods. following code show why we can remove it:
```cpp
class foo {
public:
    const int* g() const { return &n; } // returns const pointer
    int* g() { return &n; }  // ok, non-const version
    int f() const { return n; } // returns copy of n
    int f() { return n; } // meaningless. const version can do all the job
private:
    int n;
};
```

**Comment by edgarriba** [comment]
agree with removing const methods but compiler tells that are needed in:

    std::vector<const vec_t*> get_weights() const

**Comment by edgarriba** [comment]
@nyanp I've come up with this solution in order to compile but I got a segfault in testing

**Comment by nyanp** [comment]
@edgarriba 
It seems that the solution itself is ok. I'll check it after merging :)

**Comment by nyanp** [comment]
@edgarriba 
I've fixed! :+1:  0af1c04

**Comment by edgarriba** [comment]
@nyanp I still get a segfault at tests :cry:

**Comment by nyanp** [comment]
@edgarriba 
Hmm... could you show me the stack trace?

**Comment by edgarriba** [comment]
@nyanp Is this okay for you?

Type "apropos word" to search for commands related to "word"...
Reading symbols from ./tiny_cnn_test...(no debugging symbols found)...done.
(gdb) start
Temporary breakpoint 1 at 0x5b3750
Starting program: /home/eriba/software/tiny-cnn/build/tiny_cnn_test 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Temporary breakpoint 1, 0x00000000005b3750 in main ()

**Comment by nyanp** [comment]
@edgarriba
Thanks! But unfortunately we can't find where the segv occurs…
- please build the test on debug mode
- type ```backtrace``` command when you get segv in gdb


reference: http://www.unknownroad.com/rtfm/gdbtut/gdbsegfault.html

**Comment by edgarriba** [comment]
@nyanp https://github.com/nyanp/tiny-cnn/pull/157/files#r65020834 solves part of the problem, but now something weird happens with the conv layer. I got the following error after loading the net:

eriba@CVC224:~/software/tiny-cnn/build$ ./example_caffe_converter /home/eriba/software/caffe/models/bvlc_reference_caffenet/deploy.prototxt /home/eriba/software/caffe/models/bvlc_reference_caffenet/bvlc_reference_caffenet.caffemodel /home/eriba/software/bmva/networks/caffe/alexnet/imagenet_mean.binaryproto /home/eriba/software/bmva/networks/caffe/alexnet/synset_words.txt /home/eriba/Desktop/14640188712731726812104.jpg
convert Convolution => N8tiny_cnn19convolutional_layerINS_10activation8identityEEE
 shape:55x55x96
convert ReLU => N8tiny_cnn12linear_layerINS_10activation4reluEEE
 shape:55x55x96
convert Pooling => N8tiny_cnn17max_pooling_layerINS_10activation8identityEEE
 shape:27x27x96
convert LRN => N8tiny_cnn9lrn_layerINS_10activation8identityEEE
 shape:27x27x96
convert Convolution => N8tiny_cnn19convolutional_layerINS_10activation8identityEEE
 shape:27x27x256
convert ReLU => N8tiny_cnn12linear_layerINS_10activation4reluEEE
 shape:27x27x256
convert Pooling => N8tiny_cnn17max_pooling_layerINS_10activation8identityEEE
 shape:13x13x256
convert LRN => N8tiny_cnn9lrn_layerINS_10activation8identityEEE
 shape:13x13x256
convert Convolution => N8tiny_cnn19convolutional_layerINS_10activation8identityEEE
 shape:13x13x384
convert ReLU => N8tiny_cnn12linear_layerINS_10activation4reluEEE
 shape:13x13x384
convert Convolution => N8tiny_cnn19convolutional_layerINS_10activation8identityEEE
 shape:13x13x384
convert ReLU => N8tiny_cnn12linear_layerINS_10activation4reluEEE
 shape:13x13x384
convert Convolution => N8tiny_cnn19convolutional_layerINS_10activation8identityEEE
 shape:13x13x256
convert ReLU => N8tiny_cnn12linear_layerINS_10activation4reluEEE
 shape:13x13x256
convert Pooling => N8tiny_cnn17max_pooling_layerINS_10activation8identityEEE
 shape:6x6x256
convert InnerProduct => N8tiny_cnn21fully_connected_layerINS_10activation8identityEEE
 shape:4096x1x1
convert ReLU => N8tiny_cnn12linear_layerINS_10activation4reluEEE
 shape:4096x1x1
convert Dropout => N8tiny_cnn13dropout_layerE
 shape:4096x1x1
convert InnerProduct => N8tiny_cnn21fully_connected_layerINS_10activation8identityEEE
 shape:4096x1x1
convert ReLU => N8tiny_cnn12linear_layerINS_10activation4reluEEE
 shape:4096x1x1
convert Dropout => N8tiny_cnn13dropout_layerE
 shape:4096x1x1
convert InnerProduct => N8tiny_cnn21fully_connected_layerINS_10activation8identityEEE
 shape:1000x1x1
convert Softmax => N8tiny_cnn12linear_layerINS_10activation7softmaxEEE
 shape:1000x1x1
invalid group size

**Comment by nyanp** [comment]
@edgarriba 
2a6d305 may fix this problem :) but the test is still failing...

**Comment by edgarriba** [comment]
@nyanp got the following backtrace:

Program received signal SIGSEGV, Segmentation fault.
0x0000000000971918 in tiny_cnn::edge::get_data (this=0x0, worker_index=0)
    at /home/eriba/software/tiny-cnn/tiny_cnn/node.h:125
125	        return worker_specific_data_ ? &data_[worker_index] : &data_[0];
(gdb) backtrace 
#0  0x0000000000971918 in tiny_cnn::edge::get_data (this=0x0, worker_index=0)
    at /home/eriba/software/tiny-cnn/tiny_cnn/node.h:125
#1  0x0000000000972e38 in tiny_cnn::layer::forward (this=0x64d9c68, 
    worker_index=0) at /home/eriba/software/tiny-cnn/tiny_cnn/layers/layer.h:381
#2  0x000000000097511f in tiny_cnn::sequential::forward (this=0xded4a0, 
    first=std::vector of length 1, capacity 1 = {...}, worker_index=0)
    at /home/eriba/software/tiny-cnn/tiny_cnn/nodes.h:223
#3  0x0000000000996ba2 in tiny_cnn::network<tiny_cnn::sequential>::fprop (
    this=0xded498, in=std::vector of length 154587, capacity 154587 = {...}, 
    idx=0) at /home/eriba/software/tiny-cnn/tiny_cnn/network.h:665
#4  0x000000000098f08b in tiny_cnn::network<tiny_cnn::sequential>::predict (
    this=0xded498, in=std::vector of length 154587, capacity 154587 = {...})
    at /home/eriba/software/tiny-cnn/tiny_cnn/network.h:156
#5  0x0000000000986734 in tiny_cnn::network<tiny_cnn::sequential>::predict<std::vector<double, std::allocator<double> > > (this=0xded498, 
    in=std::vector of length 154587, capacity 154587 = {...})
    at /home/eriba/software/tiny-cnn/tiny_cnn/network.h:186
#6  0x000000000096ec36 in test

**Pull request merged**